README
Audio::LADSPA
==============

This is a set of extensions to host LADSPA plugins,
you can query them, or apply them to audio data.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

Or, if you prefer the CPAN installer, type:

   perl -MCPAN -e'install Audio::LADSPA'

Which will fetch, make and install the module for you.

DEPENDENCIES

LADSPA dependencies:

 The tests specifically need the demo plugins from the
 ladspa_sdk package, with LADSPA_PATH pointing to the
 directory where they are installed.

 You do not have to set the LADSPA_PATH environment variable
 if your LADSPA plugins are located in either /usr/lib/ladspa
 or /usr/local/lib/ladspa, but you will be warned when the
 modules are loaded. You'll probably want to set it anyway,
 because most other LADSPA hosts need it.

 You can download the sdk from from http://www.ladspa.org/

 Alternatively, you can skip the tests, but these modules
 are of no use without at least SOME plugins installed.

CPAN dependencies:

Also, this extension requires the following Perl modules;

  Test::More, Audio::Play, Scalar::Util and Graph::Directed
   
Install them first from CPAN. Alternatively, if you're
installing using the CPAN or CPANPLUS shell, this can be taken
care of automatically.




COPYRIGHT AND LICENCE

Copyright (C) 2003 Joost Diepenmaat

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

See the COPYING file for more information



CHANGES
0.020 -
 * Testing if the path used to find plugins is actually contains
   readable directories should work better now.
 * Added const char* to typemap, which may or may not fix the
   problems building these modules on 5.6.x perls.

0.019 -
 * Fixed the tests - there was a dependency on a LADSPA plugin
   that isn't part of the ladspa dev 

0.018 -
 * Networks are now subclasses of Class::Publisher instead of
   Class::Observable.

0.017 

 * NOT RELEASED due to bugs in Class::Observable, see
   http://rt.cpan.org/Public/Bug/Display.html?id=19507
 * Audio::LADSPA::Network is now a subclass of Class::Observable 
 * Added sample_rate and buffer_size methods to Audio::LADSPA::Network
 * Audio::LADSPA will now only load the first version of any given
   library it finds in $ENV{LADSPA_PATH}. (this used to throw a 
   pretty useless exception).
 * Upgraded ppport.h to the version shipped with perl 5.8.8

0.016 -
 Some changes to improve automatic builds and tests:
 * Default build now skips tests if the SDK libraries can't be
   found
 * Default build skips playing audio if Audio::Play isn't 
   installed
 * Documentation updates.

0.015 -
 * Audio::LADSPA::Network now works with new (0.5) Graph module
 * Removed user querying at end of tests so the tests can be run
   without user intervention
 * Some additions to Audio::LADSPA::UserGuide

0.014 -
 * Network now checks wether added plugins are already connected
   (the documentation already claimed that it did)
 * *_uniqid methods added (requires Data::Uniqid)
 * Some documentation additions

0.013 -
 * Changes to the Audio::LADSPA documentation
 * Audio::LADSPA->plugins now also returns any loaded
   perl-based ladspa plugins

0.012 -
 * Fixed bug in Network code that would trip up Network->disconnect
   2 times out of 3

0.011 -
 * added port2index method to plugins
 * added connections method to Network
 * added delete method to Network
 * allowing connections from audio port to control port
 * added documentation for all the port information methods,
   slight other improvements to the documentation for
   Audio::LADSPA::Plugin
 * added Audio::LADSPA::Plugin::Sequencer4

0.010 -
 * Oops: added Scalar::Util to the prerequisites in the top-level 
   Makefile.PL
 * Now tested with gcc-2.95.4 20011002 (Debian prerelease)
   this seems to catch the non-C'99 incompatibilities.
 * Documentation fixes (mainly typos)

0.009 -
 * All dependencies are now declared in the top-level Makefile.PL
   (should make CPAN a bit happier)
 * Some bugs have been fixed
 * Implemented a non-C'99 compatibility patch by Mike Castle.

0.008 -
 * Added Audio::LADSPA::UserGuide and expanded the documentation
   in general.
 * Added Audio::LADSPA::Plugin::Perl & Audio::LADSPA::Plugin::Play
 * made Audio::LADSPA::Network a lot more usable.

0.007 -
 * Removed Port objects - access is now done via Plugin methods
 * Made test rely in Audio::Play instead of Audio::Wav
 
0.006 -
 * Added simple buffer arithmatic
 * Fixed some bugs
 * Added simpler interface to buffer get/set methods 

0.005 - 
 * Initial release
 * Tests make 'beep beep' sound :-)
 



Makefile.PL
#!/usr/bin/perl -w
use strict;

our $VERSION = 0.019;

warn <<GPL unless -f 'Makefile';
----------------------------------------------------------------
Audio::LADSPA v$VERSION
A collection of perl modules for interfacing with LADSPA plugins
Copyright (C) 2003 - 2007 Joost Diepenmaat.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

See the COPYING file for more information.
----------------------------------------------------------------
GPL
use ExtUtils::MakeMaker;
use Config;


$ENV{LADSPA_PATH} ||= "/usr/local/lib/ladspa:/usr/lib/ladspa";


my $found = 0;
for (split/:/,$ENV{LADSPA_PATH}) {
    if (-f "$_/delay.$Config{so}") {
        $found = 1;
        last;
    }
}

unless ($found) {
	my $continue = prompt(
qq(the ladspa_sdk plugin "delay.$Config{so}" cannot be found in
LADSPA_PATH: "$ENV{LADSPA_PATH}"

The tests specifically need the plugins from the ladspa_sdk
package, with the LADSPA_PATH environment variable pointing 
to the directory where they are installed.

Please correct your LADSPA_PATH and/or download and  install
the SDK from from http://www.ladspa.org/

If you want, you can proceed without the ladspa_sdk package
installed, but most of the tests will  be skipped.

Do you want to proceed? y/n),"y");
    if ($continue !~ /^y/i) {
        exit;
    }
}

my $play_audio = 1;
if (! eval { require Audio::Play; } ) {
    $play_audio = prompt(
"No Audio::Play module was found on your system. Without it, 
we can't do audio output. If you want to test audio output,
answer yes to the following question. Audio::Play will then 
be added to the required module list. This means it will be
installed automatically if you're using the CPAN / CPANPLUS
modules.

Do you want to test audio output (requires Audio::Play)? y/n","n");
}
$play_audio = $play_audio =~ /^y/i ? 1 : 0;

WriteMakefile(
    'NAME'		=> 'Audio::LADSPA',
    'VERSION_FROM'	=> 'LADSPA.pm', # finds $VERSION
    'PREREQ_PM'		=> {
	'Test::More' => 0, 
    'Class::Publisher' => 0,
    ( $play_audio ?	('Audio::Play' => 1.000, 'Audio::Data' => 1.000) : ()),
	'Graph' => 0.5,         # NEW interface for graph
	'Scalar::Util' => 0, 
	'Data::Uniqid' => 0,
    }, # e.g., Module::Name => 1.i
    
    ($] >= 5.005 ?    ## Add these new keywords supported since 5.005
       (AUTHOR     => 'Joost Diepenmaat <jdiepen AT cpan.org>') : ()),
    'EXE_FILES'		=> ['eg/pluginfo'],
    'OPTIMIZE'		=> '-O',
);

